onnxruntime/test/providers/qnn/gemm_op_test.cc

```diff
@@ -177,6 +177,47 @@ TEST_F(QnnCPUBackendTests, Gemm_Broadcast_Bias_DynamicA_StaticB_StaticC) {
                           ExpectedEPNodeAssignment::All);
 }
 
+namespace {
+GetTestModelFn BuildReshapeGemmTestCase(const TestInputDef<float>& input, const TestInputDef<int64_t>& shape,
+                                        const TestInputDef<float>& weight, const TestInputDef<float>& bias) {
+  return [&](ModelTestBuilder& builder) {
+    std::vector<NodeArg*> reshape_inputs = {MakeTestInput<float>(builder, input),
+                                            MakeTestInput<int64_t>(builder, shape)};
+    auto* reshape_output = builder.MakeIntermediate();
+    builder.AddNode("Reshape", reshape_inputs, {reshape_output});
+    NodeArg* output = builder.MakeOutput();
+    std::vector<NodeArg*> gemm_inputs = {reshape_output, MakeTestInput<float>(builder, weight),
+                                         MakeTestInput<float>(builder, bias)};
+    builder.AddNode("Gemm", gemm_inputs, {output});
+  };
+}
+
+void RunCPUReshapeGemmTest(const TestInputDef<float>& input, const TestInputDef<int64_t>& shape,
+                           const TestInputDef<float>& weight, const TestInputDef<float>& bias,
+                           ExpectedEPNodeAssignment expected_ep_assignment, float fp32_abs_err = 1e-5f) {
+  ProviderOptions provider_options;
+
+#if defined(_WIN32)
+  provider_options["backend_path"] = "QnnCpu.dll";
+#else
+  provider_options["backend_path"] = "libQnnCpu.so";
+#endif
+  auto build_fn = BuildReshapeGemmTestCase(input, shape, weight, bias);
+  RunQnnModelTest(build_fn, provider_options, 18, expected_ep_assignment, fp32_abs_err);
+}
+
+}  // namespace
+
+TEST_F(QnnCPUBackendTests, ReshapeGemmFusion) {
+  std::vector<float> input_data = {1.0f, 2.0f, 3.0f, 4.0f, -1.0f, -2.0f, -3.0f, -4.0f};
+  std::vector<int64_t> shape_data = {4, 2};
+  std::vector<float> weight_data(6, 1.0f);
+  std::vector<float> bias_data = {1.0f, 2.0f, 3.0f};
+  RunCPUReshapeGemmTest(TestInputDef<float>({2, 2, 2}, false, input_data), TestInputDef<int64_t>({2}, true, shape_data),
+                        TestInputDef<float>({2, 3}, true, weight_data), TestInputDef<float>({3}, true, bias_data),
+                        ExpectedEPNodeAssignment::All);
```

Comment:
What validates that the fusion took place? Naively I would have expected all nodes to be assigned to QNN either way, and that the fusion results in better performance, so I wouldn't have expected the ExpectedEPNodeAssignment::All check to be sufficient.

Reply:
I actually realized this problem at the beginning, but don't know how to solve it. The DQ->Conv->Act->Q also tested with this way, not able to check the fusion. The HardSigmoid->Mul fusing to HardSwish just does not have any test covered. Adrian has a debug branch to dump QNN EP graph to JSON file, it's better he push his branch to main, so we can dump the QNN EP JSON file (by setting session configs) and check.